data/util/NumericShaperUtility.js

```diff
@@ -0,0 +1,294 @@
+					//		Allowed values:
+					//		1. "ltr"
+					//		2. "rtl"
+					//		3. "auto"
+					// returns:
+					//		The shaped string.
+					
+					if(!text){
+						return;
+					}
+					
+					if(["National", "Nominal", "Contextual"].indexOf(shaperType) === -1){
+						return text;
+					}
+					
+					if(shaperType === "National"){
+						return this._shapeArabic(text);
+					}else if(shaperType === "Nominal"){
+						return this._shapeEuropean(text);
+					}else if(shaperType === "Contextual" && textDir === "rtl"){
+						this._context = 2;
+						return this._shapeContextual(text);
+					}else{
+						this._context = 1;
+						return this._shapeContextual(text);
+					}
+				},
+				
+				_shapeEuropean: function (/*String*/ text){
+					// summary:
+					//		Converts the digits in the text to European digits.
+					// text: String
+					//		The text to be shaped.
+					// return:
+					//		The shaped string in European format.
+					// tags:
+			        //		private
+					
+					var textArr = text.split("");
+					var c = "";
+					for (var i = 0; i < text.length; ++i) {
+						c = textArr[i];
+						if (c >= "\u0660" && c <= "\u0669") { // Arabic digits range
+							textArr[i] = c.charCodeAt(0) - 1632;
+						}
+					}
+					return textArr.join("");
+				},
+				
+				_shapeArabic: function(/*String*/ text){
+					// summary:
+					//		Converts the digits in the text to Arabic digits.
+					// text: String
+					//		The text to be shaped.
+					// return:
+					//		The shaped string in Arabic format.
+					// tags:
+			        //		private
+					
+					var textArr = text.split("");
+					var c = "";
+					for (var i = 0; i < text.length; ++i) {
+						c = textArr[i];
+						if (c >= "0" && c <= "9") { // European digits range
+							textArr[i] = String.fromCharCode(parseInt(c) + 1632);
+						}
+					}
+					return textArr.join("");
+				},
+				
+				_shapeContextual: function(/*String*/ text){
+					// summary:
+					//		Converts the digits in the text to European or Arabic digits
+					//		According to the type of the preceding strong character.
+					// text: String
+					//		The text to be shaped.
+					// return:
+					//		The shaped string.
+					// tags:
+			        //		private
+					
+					var textArr = text.split("");
+					var c = "";
+					for (var i = 0; i < text.length; ++i) {
+						c = textArr[i];
+						if (c >= "0" && c <= "9") {  // European digits range
+							if(this._context === 2){ // context = 2 so convert from European to Arabic. else skip.
+								textArr[i] = String.fromCharCode(parseInt(c) + 1632);
+							}
+						}else if (c >= "\u0660" && c <= "\u0669"){ // Arabic digits range
+							if(this._context === 1){ // context = 1 so convert from Arabic to European. else skip.
+								textArr[i] = c.charCodeAt(0) - 1632;
+							}
+						}else{ // this is not an Arabic or European digit.
+							this._setContext(c);
+						}
+						
+					}
+					return textArr.join("");
+				},
```

Comment:
It seems wrong that `this._context` persists across calls to `shape()`, because it means that one call to `shape()` can affect the behavior of the next call to `shape()`.